presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java

```diff
@@ -826,6 +838,82 @@ private Global deriveGlobalProperties(TableLayout layout, Map<ColumnHandle, Vari
         return inputToOutput;
     }
 
+    private static Optional<RowExpression> rewriteToCoalesce(
+            TypeManager typeManager,
+            ImmutableBiMap<VariableReferenceExpression, VariableReferenceExpression> joinMap,
+            RowExpression column)
+    {
+        checkArgument(column instanceof VariableReferenceExpression || column instanceof ConstantExpression, "Expect input partition to be variables or constants");
+        if (column instanceof VariableReferenceExpression) {
+            VariableReferenceExpression variable = (VariableReferenceExpression) column;
+            ImmutableBiMap map;
+            if (!joinMap.containsKey(variable) && !joinMap.containsValue(variable)) {
+                return Optional.empty();
+            }
+            if (joinMap.containsKey(variable)) {
+                map = joinMap;
+            }
+            else {
+                map = joinMap.inverse();
+            }
+            return Optional.of(new SpecialFormExpression(
+                    COALESCE,
+                    typeManager.getCommonSuperType(variable.getType(), ((VariableReferenceExpression) map.get(variable)).getType()).get(),
+                    ImmutableList.of(column, (RowExpression) map.get(variable))));
+        }
+        else {
+            return Optional.of(column);
+        }
+    }
+
+    private static Optional<RowExpression> rewriteExpression(
+            Map<VariableReferenceExpression, RowExpression> assignments,
+            TypeProvider types,
+            Map<VariableReferenceExpression, VariableReferenceExpression> identities,
```

Comment:
If all `assignments` are passed, does it still make sense to pass the `identities` as well? It should be pretty straightforward to check if the assignment is an identity within this method.

Reply:
`identities` is a reverse map of `assignments` for identity assignments. It's more efficient to calculate this reverse map before translating each property with `rewriteExpression`. Another way to do this is to pass a reversed `assignment` (`Map<RowExpression, VariableReferenceExpression>`) to `rewriteExpression`. What do you think?

Reply:
My worry about reversing the assignment is whether there would be cases that the same expression is assigned to two different variables, like
```
a = foo(x)
b = foo(x)
```
So I prefer to keep the logic the way it is.